# Lesson 17: `pandas` Part 7a: Combining Datasets with `concat()`

# Learning Objectives
## By the end of this tutorial you will be able to:
1. Combine DataFrames and/or Series with `concat()`
2. Understand a multi-index
3. Reset an index with `reset_index()`
4. Perform descriptive analytics on a combined DataFrame

## Files Needed for this lesson:
>- [`CAvideos.csv`](https://drive.google.com/file/d/1nLm7qD-StcFX5e7nY2eWBRd0gueeKPTb/view?usp=sharing)
>- [`GBvideos.csv`](https://drive.google.com/file/d/1tUVpOtlcg273hPXVhnyPDOES_8LL9Xxg/view?usp=sharing)
>- Download the csv files move to your working directory

## The general steps to working with pandas:
1. import pandas as pd
2. Create or load data into a pandas DataFrame or Series
3. Reading data with `pd.read_`
>- Excel files: `pd.read_excel('fileName.xlsx')`
>- Csv files: `pd.read_csv('fileName.csv')`
>- Note: if the file you want to read into your notebook is not in the same folder you can do one of two things:
>>- Move the file you want to read into the same folder/directory as the notebook
>>- Type out the full path into the read function
4. After steps 1-3 you will want to check out your DataFrame
>- Use `shape` to see how many records and columns are in your DataFrame
>- Use `head()` to show the first 5-10 records in your DataFrame

# Introduction Notes on Combining Data Using `pandas`
1. Being able to combine data from multiple sources is a critical skill for analytics professionals
2. We will learn the `pandas` way of combining data but there are similarities here to SQL
3. Why combine data with `pandas` if you can do the same thing in SQL?
>- The answer to this depends on the project
>- Some projects may be completed more efficiently all with `pandas` so you wouldn't necessarily need SQL
>- For some projects incorporating SQL into our python code makes sense
>- In a an analytics job, you will likely use both python and SQL to get the job done!

# Initial set-up steps
1. import modules and check working directory
2. Read data in
3. Check the data

In [1]:
# Imports

import os, pandas as pd

from google.colab import drive

drive.mount('/content/drive/')

os.chdir('/content/drive/MyDrive/Files_for_pandas/')



Mounted at /content/drive/


In [2]:
# change to desired directory

os.listdir()

['winemag-data-130k-v2.csv',
 'wine.csv',
 'students.csv',
 'students100.xlsx',
 'GBvideos.csv',
 'CAvideos.csv',
 'DJIA.csv',
 'AAPL.csv',
 'vendor.csv',
 'customer.csv',
 'invoice.csv',
 'line.csv',
 'product.csv',
 'q2_3chart.png',
 'q2_4chart.png',
 '3.2Annual_ETF_Returns.png',
 '3.4BoxPlot.png',
 '3.6regplot.png',
 '3.7pairwisePlot.png',
 'lesson18_4.1_scatter.png',
 'Ecommerce.csv',
 'MovieSurvey.csv',
 'Untitled Diagram.drawio',
 'Smarket.csv',
 'Caravan.csv']

# Step 2 Read Data Into a DataFrame with `read_csv()`
Make sure to download and store the following files in your working directory.

>- file names:
>>- `CAvideos.csv`
>>- `GBvideos.csv`

In [3]:
can_youtube =  pd.read_csv('CAvideos.csv')

uk_youtube = pd.read_csv('GBvideos.csv')

### Check how many rows and columns are in our DataFrames

In [4]:
print(can_youtube.shape)

print(uk_youtube.shape)



(40881, 16)
(38916, 16)


### Check a couple of rows of data in one of the new DataFrames

In [9]:
can_youtube.head(2)

video_id trending_date                                       title  \
0  n1WpP7iowLc      17.14.11  Eminem - Walk On Water (Audio) ft. Beyoncé   
1  0dBIkQ4Mz1M      17.14.11               PLUSH - Bad Unboxing Fan Mail   

  channel_title  category_id              publish_time  \
0    EminemVEVO           10  2017-11-10T17:00:03.000Z   
1     iDubbbzTV           23  2017-11-13T17:00:00.000Z   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...

In [10]:
uk_youtube.head(2)

video_id trending_date                                          title  \
0  Jw1Y-zhQURU      17.14.11  John Lewis Christmas Ad 2017 - #MozTheMonster   
1  3s1rvMFUweQ      17.14.11      Taylor Swift: …Ready for It? (Live) - SNL   

         channel_title  category_id              publish_time  \
0           John Lewis           26  2017-11-10T07:38:29.000Z   
1  Saturday Night Live           24  2017-11-12T06:24:44.000Z   

                                                tags    views  likes  \
0  christmas|"john lewis christmas"|"john lewis"|...  7224515  55681   
1  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...  1053632  25561   

   dislikes  comment_count                                  thumbnail_link  \
0     10247           9479  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
1      2294           2757  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   

                                         description  
0  Click here to continue the story and make your...  
1  Musical guest Taylor Swift performs …Ready for...

## Check the datatypes

In [6]:
can_youtube.dtypes

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
dtype: object

In [7]:
uk_youtube.dtypes

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
dtype: object

# Combining DataFrames
>- The three common ways to combine datastest in pandas is with `concat()`, `join()`, and `merge()`
>- `concat()` will take two DataFrames or Series and append them together
>>- This is basically taking DataFrames and stacking their data on top of each other into one DataFrame
>>- For `concat()` you need the columns/fields in both DataFrames to the be the same
>- `join()` "links" DataFrames together based on a common field/column between the two
>- `merge()` also links DataFrames together based on common field/columns but with different syntax.
>>- We will cover the most basic join in this class
>>- A more in depth study of joins is provided in SQL focused courses
>>- Pandas join reference for further study: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html


# Using the YouTube DataFrames to practice combining data with pandas
>- The YouTube datasets store data on various YouTube trending statistics
>- Our example datasets show several months of data and daily trending YouTube videos.

>- For more information and for other YouTube datasets see the following link:
>>- https://www.kaggle.com/datasnaek/youtube-new

### First, creating a new DataFrame that appends the Canadian and British YouTube DataFrames

In [13]:
CanUk = pd.concat([can_youtube, uk_youtube],
                  keys = ['can','uk'],
                  names = ['country', 'rowid_country']
                  )

In [ ]:
CanUk

#### Some notes on the previous code
>- Line 1: We define a new DataFrame named `CanUK` which is defined as the concatenation,`concat()`, of two datasets
>>- Dataset 1 = canadian_youtube
>>- Dataset 2 = uk_youtube
>>- The `concat()` function takes the two (or more if applicable) DataFrames and "stacks" them on top of each
>- Line 2: We use `keys` option to define a multi-index (aka hierarchical index)
>>- Because our datasets represent YouTube videos from different countries we pass the abbreviated names of those countries as a list to `keys`
>>- Enter the keys names in order they appear in line 1 (e.g., 'can' first, 'uk' second)
>- Line 3: We use the `names` option to label our index columns from line 2
>>- Without the `names` option we would not have anything above our index columns

### Check the index for any dataframe using `DataFrame.index`
>- Note how `concat()` uses the rowid's for each country's dataset versus continuing the count

In [15]:
CanUk.index

MultiIndex([('can',     0),
            ('can',     1),
            ('can',     2),
            ('can',     3),
            ('can',     4),
            ('can',     5),
            ('can',     6),
            ('can',     7),
            ('can',     8),
            ('can',     9),
            ...
            ( 'uk', 38906),
            ( 'uk', 38907),
            ( 'uk', 38908),
            ( 'uk', 38909),
            ( 'uk', 38910),
            ( 'uk', 38911),
            ( 'uk', 38912),
            ( 'uk', 38913),
            ( 'uk', 38914),
            ( 'uk', 38915)],
           names=['country', 'rowid_country'], length=79797)

### Take a look at our new DataFrame

#### Did using `concat()` work to append the two DataFrames together?
>- Check the shape of your new DataFrame
>- Compare the number of records to each one individually
>>- canadian_youtube = 40881 records
>>- uk_youtube = 38916 records
>>- 40881 + 38916 = 79797 total records

#### `reset_index`:
##### Note: You can reset a an index with `reset_index`
>- This can be useful for some situations
>- For a multi-index you can pass the `level` option and specify what index you want to reset
>- Note: To make the change to our current DataFrame we would need to use the option, `inplace=True`

In [16]:
CanUk.reset_index(level = 'country').head(2)

country     video_id trending_date  \
rowid_country                                      
0                 can  n1WpP7iowLc      17.14.11   
1                 can  0dBIkQ4Mz1M      17.14.11   

                                                    title channel_title  \
rowid_country                                                             
0              Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                           PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   

               category_id              publish_time  \
rowid_country                                          
0                       10  2017-11-10T17:00:03.000Z   
1                       23  2017-11-13T17:00:00.000Z   

                                                            tags     views  \
rowid_country                                                                
0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   
1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   

                likes  dislikes  comment_count  \
rowid_country                                    
0              787425     43420         125882   
1              127794      1688          13030   

                                               thumbnail_link  \
rowid_country                                                   
0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

               comments_disabled  ratings_disabled  video_error_or_removed  \
rowid_country                                                                
0                          False             False                   False   
1                          False             False                   False   

                                                     description  
rowid_country                                                     
0              Eminem's new track Walk on Water ft. Beyoncé i...  
1              STill got a lot of packages. Probably will las...

---
# End of Video 1
---

# Now some descriptive analytics

### What channels have the most trending videos?

In [17]:
CanUk['channel_title'].value_counts()

The Late Show with Stephen Colbert        361
TheEllenShow                              357
Jimmy Kimmel Live                         351
The Tonight Show Starring Jimmy Fallon    345
Breakfast Club Power 105.1 FM             342
                                         ... 
SABRE NORRIS & The Norris Nuts              1
SaggyJ                                      1
ItsSkyLOL                                   1
Punjab 24hr                                 1
The Chainsmokers                            1
Name: channel_title, Length: 5956, dtype: int64

### What are the quantitative descriptive statistics for TheEllenShow?

In [18]:
CanUk.where(CanUk['channel_title']== 'TheEllenShow').describe()

category_id         views          likes      dislikes  comment_count
count        357.0  3.570000e+02     357.000000    357.000000     357.000000
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000

##### Alternatively, you can use `loc[]` to peform the filtering operation
>- The use of `where()` or `loc[]` depends on the question/purpose or sometimes just personal preference

In [19]:
CanUk.loc[CanUk['channel_title']== 'TheEllenShow'].describe()

category_id         views          likes      dislikes  comment_count
count        357.0  3.570000e+02     357.000000    357.000000     357.000000
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000

### What were the total YouTube videos, total views, likes and dislikes for TheEllenShow?
>- Using the agg() function to calculate specific aggregations on different columns

In [22]:
CanUk.where(CanUk['channel_title']=='TheEllenShow').agg({'title': ['count'],
                                                         'views': ['sum'],
                                                         'likes': ['sum'],
                                                         'dislikes': ['sum']})

title        views       likes  dislikes
count  357.0          NaN         NaN       NaN
sum      NaN  635395636.0  16514024.0  474288.0

## What are the totals for TheEllenShow's top 5 most viewed videos?
>- Only include the title names as part of the output (not channel or any other categorical fields)
>- Include total views, likes, dislikes, and comment count in the output

In [24]:
CanUk[['title','views','likes','dislikes','comment_count']].where(CanUk['channel_title']=='TheEllenShow').\
groupby(['title']).sum().sort_values(by= 'views', ascending = False).head(5)

views      likes  \
title                                                                       
Jennifer Lawrence Explains Her Drunk Alter Ego ...  71621685.0  1302161.0   
Kid Yodeler Mason Ramsey Performs                   68553312.0  2697959.0   
Kim Kardashian Lets Gender of Third Child Slip      52985123.0   776177.0   
Billionaire Bill Gates Guesses Grocery Store Pr...  37076353.0   542804.0   
Laurel or Yanny?                                    28695358.0   613487.0   

                                                    dislikes  comment_count  
title                                                                        
Jennifer Lawrence Explains Her Drunk Alter Ego ...   35255.0        49762.0  
Kid Yodeler Mason Ramsey Performs                   144604.0            0.0  
Kim Kardashian Lets Gender of Third Child Slip       26509.0        50940.0  
Billionaire Bill Gates Guesses Grocery Store Pr...   29036.0        42720.0  
Laurel or Yanny?                                     18568.0       100731.0

# Some Notes on the Previous Example
>- Our pandas code in the previous example is similar to SQL in the following ways
    1. `loc[CanUk.channel_title == 'TheEllenShow',` is SQL equivalent to `WHERE channel_title = 'TheEllenShow'`
    2. `['title','views','likes','dislikes','comment_count']` is SQL equivalent to:
        `SELECT title, sum(views),sum(likes),sum(dislikes),sum(comment_count)`
    3. `groupby(['title`]) is SQL equivalent to GROUP BY title
    4. Now in pandas we enter the aggregation after the `groupby()`, in this example `sum()`
      >>- In SQL we write the aggregation in the SELECT statement
      
## In future lessons we will continue to learn how pandas and SQL relate